### PR TITLE
HUD traits now apply their corresponding hud automatically. Most clothing/item/etc sources of HUDs now only use traits

### PR DESCRIPTION
**Pull request by SmArtKar**
## About The Pull Request

Currently if you want to apply a HUD you usually add both its trait and the HUD itself. Only exceptions are things like simplemobs where you should avoid adding the hud trait since it adds security/med DB access and such, but there is no cases where you'd want to apply the trait and not apply the hud.

Requested by Melbert about a week ago.
image

## Why It's Good For The Game

This makes working with HUDs significantly easier, as you no longer have to bother with manually adding/removing them. Also potentially removes an edge case where if your hud could get removed while keeping the trait.

## Changelog
:cl:
refactor: HUD traits now apply their corresponding hud automatically
/:cl:

### CODE REVIEW

File: code/modules/clothing/glasses/hud.dm [hunk 1]
```diff
@@ -118,7 +96,6 @@
 	name = "diagnostic HUD"
 	desc = "A heads-up display capable of analyzing the integrity and status of robotics and exosuits."
 	icon_state = "diagnostichud"
-	hud_types = list(DATA_HUD_DIAGNOSTIC_BASIC)
 	clothing_traits = list(TRAIT_DIAGNOSTIC_HUD)
```
Comment (MrMelbert) [inline code/modules/clothing/glasses/hud.dm:99]: Diagnostic huds, currently, give `DATA_HUD_DIAGNOSTIC_BASIC`
In this PR, `TRAIT_DIAGNOSTIC_HUD` gives `DATA_HUD_DIAGNOSTIC_ADVANCED`
So this is an error. (The only error I noticed in the PR though) 

But the only difference between the two is the advanced hud shows but paths. Maybe we should just separate bot paths into its own hud (that things like the AI and observers use)?

**Review by MrMelbert — CHANGES_REQUESTED** [review]

File: code/modules/mob/dead/observer/observer.dm [hunk 2]
```diff
@@ -757,14 +755,14 @@ This is the proc mobs get to turn into a ghost. Forked from ghostize due to comp
 	return
 
 /mob/dead/observer/proc/show_data_huds()
-	for(var/hudtype in datahuds)
-		var/datum/atom_hud/data_hud = GLOB.huds[hudtype]
-		data_hud.show_to(src)
+	ADD_TRAIT(src, TRAIT_SECURITY_HUD, ref(src))
+	ADD_TRAIT(src, TRAIT_MEDICAL_HUD, ref(src))
+	ADD_TRAIT(src, TRAIT_DIAGNOSTIC_HUD, ref(src))
```
Comment (MrMelbert) [inline code/modules/mob/dead/observer/observer.dm:760]:
```suggestion
	ADD_TRAIT(src, TRAIT_SECURITY_HUD, REF(src))
	ADD_TRAIT(src, TRAIT_MEDICAL_HUD, REF(src))
	ADD_TRAIT(src, TRAIT_DIAGNOSTIC_HUD, REF(src))
```

File: code/modules/mob/dead/observer/observer.dm [hunk 3]
```diff
@@ -757,14 +755,14 @@ This is the proc mobs get to turn into a ghost. Forked from ghostize due to comp
...
 
 /mob/dead/observer/proc/remove_data_huds()
-	for(var/hudtype in datahuds)
-		var/datum/atom_hud/data_hud = GLOB.huds[hudtype]
-		data_hud.hide_from(src)
+	REMOVE_TRAIT(src, TRAIT_SECURITY_HUD, ref(src))
+	REMOVE_TRAIT(src, TRAIT_MEDICAL_HUD, ref(src))
+	REMOVE_TRAIT(src, TRAIT_DIAGNOSTIC_HUD, ref(src))
```
Comment (MrMelbert) [inline code/modules/mob/dead/observer/observer.dm:765]:
```suggestion
	REMOVE_TRAIT(src, TRAIT_SECURITY_HUD, REF(src))
	REMOVE_TRAIT(src, TRAIT_MEDICAL_HUD, REF(src))
	REMOVE_TRAIT(src, TRAIT_DIAGNOSTIC_HUD, REF(src))
```

File: code/modules/mob/mob.dm [hunk 4]
```diff
@@ -1598,3 +1599,21 @@
...
 	..()
 	SEND_SIGNAL(src, COMSIG_MOB_KEYDOWN, key, client, full_key)
+
+/mob/proc/setup_hud_traits()
+	RegisterSignal(src, SIGNAL_ADDTRAIT(TRAIT_SECURITY_HUD), PROC_REF(hud_trait_enabled))
+	RegisterSignal(src, SIGNAL_ADDTRAIT(TRAIT_MEDICAL_HUD), PROC_REF(hud_trait_enabled))
+	RegisterSignal(src, SIGNAL_ADDTRAIT(TRAIT_DIAGNOSTIC_HUD), PROC_REF(hud_trait_enabled))
+	RegisterSignal(src, SIGNAL_REMOVETRAIT(TRAIT_SECURITY_HUD), PROC_REF(hud_trait_disabled))
+	RegisterSignal(src, SIGNAL_REMOVETRAIT(TRAIT_MEDICAL_HUD), PROC_REF(hud_trait_disabled))
+	RegisterSignal(src, SIGNAL_REMOVETRAIT(TRAIT_DIAGNOSTIC_HUD), PROC_REF(hud_trait_disabled))
```
Comment (MrMelbert) [inline code/modules/mob/mob.dm:1609]: Can probably make this a bit easier to expand upon: 
```dm
for(var/hud_trait in GLOB.trait_to_hud)
    RegisterSignal(src, SIGNAL_ADDTRAIT(hud_trait), PROC_REF(hud_trait_enabled))
    RegisterSignal(src, SIGNAL_REMOVETRAIT(hud_trait), PROC_REF(hud_trait_enabled))
```

**Comment by MrMelbert** [comment]
Good PR

File: code/modules/mob/dead/observer/observer.dm [hunk 5]
```diff
@@ -757,14 +755,16 @@ This is the proc mobs get to turn into a ghost. Forked from ghostize due to comp
 	return
 
 /mob/dead/observer/proc/show_data_huds()
-	for(var/hudtype in datahuds)
-		var/datum/atom_hud/data_hud = GLOB.huds[hudtype]
-		data_hud.show_to(src)
+	ADD_TRAIT(src, TRAIT_SECURITY_HUD, REF(src))
+	ADD_TRAIT(src, TRAIT_MEDICAL_HUD, REF(src))
+	ADD_TRAIT(src, TRAIT_DIAGNOSTIC_HUD, REF(src))
+	ADD_TRAIT(src, TRAIT_BOT_PATH_HUD, REF(src))
```
Comment (san7890) [inline code/modules/mob/dead/observer/observer.dm:761]: use `add_traits()`, we can multiline the list to make that cleaner

File: code/modules/mob/dead/observer/observer.dm [hunk 6]
```diff
@@ -757,14 +755,16 @@ This is the proc mobs get to turn into a ghost. Forked from ghostize due to comp
...
 
 /mob/dead/observer/proc/remove_data_huds()
-	for(var/hudtype in datahuds)
-		var/datum/atom_hud/data_hud = GLOB.huds[hudtype]
-		data_hud.hide_from(src)
+	REMOVE_TRAIT(src, TRAIT_SECURITY_HUD, REF(src))
+	REMOVE_TRAIT(src, TRAIT_MEDICAL_HUD, REF(src))
+	REMOVE_TRAIT(src, TRAIT_DIAGNOSTIC_HUD, REF(src))
+	REMOVE_TRAIT(src, TRAIT_BOT_PATH_HUD, REF(src))
```
Comment (san7890) [inline code/modules/mob/dead/observer/observer.dm:767]: use `remove_traits()`. also multiline dat list

File: code/modules/surgery/organs/internal/cyberimp/augments_eyes.dm [hunk 7]
```diff
@@ -13,62 +13,51 @@
...
 
 
 /obj/item/organ/internal/cyberimp/eyes/hud/proc/toggle_hud(mob/living/carbon/eye_owner)
 	if(toggled_on)
-		if(HUD_type)
-			var/datum/atom_hud/hud = GLOB.huds[HUD_type]
-			hud.hide_from(eye_owner)
 		toggled_on = FALSE
+		for (var/HUD_trait in HUD_traits)
+			ADD_TRAIT(eye_owner, HUD_trait, ORGAN_TRAIT)
```
Comment (san7890) [inline code/modules/surgery/organs/internal/cyberimp/augments_eyes.dm:25]:
```suggestion
		eye_owner.add_traits(HUD_traits, ORGAN_TRAIT)
```

File: code/modules/surgery/organs/internal/cyberimp/augments_eyes.dm [hunk 8]
```diff
@@ -13,62 +13,51 @@
...
-	else
-		if(HUD_type)
-			var/datum/atom_hud/hud = GLOB.huds[HUD_type]
-			hud.show_to(eye_owner)
-		toggled_on = TRUE
-		balloon_alert(eye_owner, "hud enabled")
+		return
+	toggled_on = TRUE
+	for (var/HUD_trait in HUD_traits)
+		REMOVE_TRAIT(eye_owner, HUD_trait, ORGAN_TRAIT)
```
Comment (san7890) [inline code/modules/surgery/organs/internal/cyberimp/augments_eyes.dm:30]:
```suggestion
	eye_owner.remove_traits(HUD_traits, ORGAN_TRAIT)
```

File: code/modules/surgery/organs/internal/cyberimp/augments_eyes.dm [hunk 9]
```diff
@@ -13,62 +13,51 @@
...
 /obj/item/organ/internal/cyberimp/eyes/hud/Insert(mob/living/carbon/eye_owner, special = FALSE, movement_flags)
 	. = ..()
 	if(!.)
 		return
-	if(HUD_type)
-		var/datum/atom_hud/hud = GLOB.huds[HUD_type]
-		hud.show_to(eye_owner)
-	if(HUD_trait)
+	for (var/HUD_trait in HUD_traits)
 		ADD_TRAIT(eye_owner, HUD_trait, ORGAN_TRAIT)
```
Comment (san7890) [inline code/modules/surgery/organs/internal/cyberimp/augments_eyes.dm:38]: add_traits here

File: code/modules/surgery/organs/internal/cyberimp/augments_eyes.dm [hunk 10]
```diff
@@ -13,62 +13,51 @@
...
 	toggled_on = TRUE
 
 /obj/item/organ/internal/cyberimp/eyes/hud/Remove(mob/living/carbon/eye_owner, special, movement_flags)
 	. = ..()
-	if(HUD_type)
-		var/datum/atom_hud/hud = GLOB.huds[HUD_type]
-		hud.hide_from(eye_owner)
-	if(HUD_trait)
+	for (var/HUD_trait in HUD_traits)
 		REMOVE_TRAIT(eye_owner, HUD_trait, ORGAN_TRAIT)
```
Comment (san7890) [inline code/modules/surgery/organs/internal/cyberimp/augments_eyes.dm:44]: remove_traits here

**Review by san7890 — CHANGES_REQUESTED** [review]
we have `add_traits()` and `remove_traits()` procs now

**Review by Ghommie — APPROVED** [review]

**Pull request merged**